sql/core/src/main/scala/org/apache/spark/sql/SQLImplicits.scala

```diff
@@ -110,4 +121,181 @@ private[sql] abstract class SQLImplicits {
     DataFrameHolder(
       _sqlContext.internalCreateDataFrame(rows, StructType(StructField("_1", dataType) :: Nil)))
   }
+
+  /**
+   * ::Experimental::
+   *
+   * Pimp my library decorator for tungsten caching of DataFrames.
+   * @since 1.5.1
+   */
+  @Experimental
+  implicit class TungstenCache(df: DataFrame) {
+    /**
+     * Packs the rows of [[df]] into contiguous blocks of memory.
+     * @param compressionType "" (default), "lz4", "lzf", or "snappy", see
+     *                        [[CompressionCodec.ALL_COMPRESSION_CODECS]]
+     * @param blockSize size of each MemoryBlock (default = 4 MB)
+     */
+    def tungstenCache(
+        compressionType: String = "", blockSize: Int = 4000000): (RDD[_], DataFrame) = {
+      val schema = df.schema
+
+      val convert = CatalystTypeConverters.createToCatalystConverter(schema)
+      val internalRows = df.rdd.map(convert(_).asInstanceOf[InternalRow])
+      val cachedRDD = internalRows.mapPartitions { rowIterator =>
+        val bufferedRowIterator = rowIterator.buffered
```

Comment:
L166 calls `head` in order to see if we can fit this row into the current `MemoryBlock`; we want to start a new `MemoryBlock` without advancing the iterator if it doesn't fit